onnxruntime/core/providers/webnn/builders/impl/dynamicQuantizeLinear_op_builder.cc

```diff
@@ -14,34 +14,194 @@
+
+  // X_Min = ReduceMin <keepdims: int = 0> (x)
+  common_options.set("label", node.Name() + "_x_min");
+  emscripten::val x_min = model_builder.GetBuilder().call<emscripten::val>("reduceMin", input, common_options);
+
+  // X_Min_Adjusted = Min (X_Min, Q_Min)
+  common_options.set("label", node.Name() + "_x_min_adjusted");
+  emscripten::val x_min_adjusted = model_builder.GetBuilder().call<emscripten::val>("min", x_min, q_min, common_options);
+
+  // X_Max = ReduceMax <keepdims: int = 0> (x)
+  common_options.set("label", node.Name() + "_x_max");
+  emscripten::val x_max = model_builder.GetBuilder().call<emscripten::val>("reduceMax", input, common_options);
+
+  // X_Max_Adjusted = Max (X_Max, Q_Min)
+  common_options.set("label", node.Name() + "_x_max_adjusted");
+  emscripten::val x_max_adjusted = model_builder.GetBuilder().call<emscripten::val>(
+      "max", x_max, q_min, common_options);
+
+  // X_Range = Sub (X_Max_Adjusted, X_Min_Adjusted)
+  common_options.set("label", node.Name() + "_x_range");
+  emscripten::val x_range = model_builder.GetBuilder().call<emscripten::val>(
+      "sub", x_max_adjusted, x_min_adjusted, common_options);
 
-  output_array = model_builder.GetBuilder().call<emscripten::val>("dynamicQuantizeLinear", input, options);
+  // Scale = Div (X_Range, Q_Max)
+  common_options.set("label", node.Name() + "_scale");
+  emscripten::val scale = model_builder.GetBuilder().call<emscripten::val>("div", x_range, q_max, common_options);
 
-  for (size_t i = 0, count = output_array["length"].as<size_t>(); i < count; i++) {
-    model_builder.AddOperand(node.OutputDefs()[i]->Name(), std::move(output_array[i]));
+  // Min_Scaled = Div (X_Min_Adjusted, Scale)
+  common_options.set("label", node.Name() + "_min_scaled");
+  emscripten::val min_scaled = model_builder.GetBuilder().call<emscripten::val>(
+      "div", x_min_adjusted, scale, common_options);
+
+  // Initial_ZeroPoint_FP = Sub (Q_Min, Min_Scaled)
+  common_options.set("label", node.Name() + "_initial_zero_point_fp");
+  emscripten::val initial_zero_point_fp = model_builder.GetBuilder().call<emscripten::val>(
+      "sub", q_min, min_scaled, common_options);
+
+  // Clipped_ZeroPoint_FP = Clip (Initial_ZeroPoint_FP, Q_Min, Q_Max)
+  emscripten::val clip_options = emscripten::val::object();
+  clip_options.set("label", node.Name() + "_clipped_zero_point_fp");
+  clip_options.set("minValue", 0);
+  clip_options.set("maxValue", 255);
+  emscripten::val clipped_zero_point_fp = model_builder.GetBuilder().call<emscripten::val>(
+      "clamp", initial_zero_point_fp, clip_options);
+
+  // Rounded_ZeroPoint_FP = Round (Clipped_ZeroPoint_FP)
+  common_options.set("label", node.Name() + "_rounded_zero_point_fp");
+  emscripten::val rounded_zero_point_fp = model_builder.GetBuilder().call<emscripten::val>(
+      "roundEven", clipped_zero_point_fp, common_options);
+
+  // Zeropoint = Cast <to: int = 2> (Rounded_ZeroPoint_FP)
```

Comment:
```suggestion
  // ZeroPoint = Cast <to: int = 2> (Rounded_ZeroPoint_FP)
```
Nit for consistency with "Rounded_ZeroPoint_FP" (don't create a whole PR just for this :b, but if you happen to touch this function in the future...).